### PR TITLE
add firstindex check and fix a bug in mapcols!

### PR DESCRIPTION
**Pull request by bkamins**
We currently assume that vectors are 1-based in DataFrames.jl. This PR makes sure we check for that. Not checking this will lead to bugs. Also I found some small bug in `mapcols!` in the process related to `AbstractRange` handling.

I think we should decide on this before 1.0 release (it is better to be strict I think now; we can be more flexible later).

I will add tests, NEWS.md and update documentation with this PR after we agree that we want to add these changes.

### CODE REVIEW

File: src/abstractdataframe/iteration.jl [hunk 1]
```diff
@@ -454,6 +454,16 @@ function mapcols!(f::Union{Function, Type}, df::DataFrame)
...
         throw(DimensionMismatch("lengths of returned vectors must be identical"))
     end
+
+    for (i, col) in enumerate(vs)
+        fic = firstindex(col)
+        if fic != 1
+            throw(ArgumentError("Currently DataFrames.jl supports only " *
+                                "columns that use 1-based indexing and " *
+                                "column $i has starting index equal to $fic"))
+        end
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl:464]: How about moving this to a separate function to avoid repeating the error message?

Comment (bkamins) [reply src/abstractdataframe/iteration.jl:464]: OK - makes sense. Actually I think we should even remove all uses of the `_columns` function as using it leads to unsafe code. I will work on this. I understand you agree with a general direction I propose here.

**Comment by bkamins** [comment]
@nalimilan - are you OK with these proposed changes. If yes I would finalize the PR

**Comment by bkamins** [comment]
@nalimilan - the issue came up again in #2604. Are you OK to finalize this PR in the form I have proposed?

File: src/abstractdataframe/iteration.jl (hunk 1 above)
Comment (bkamins) [reply src/abstractdataframe/iteration.jl:464]: I will leave `_columns` for now as we have different use-cases for it. I have added a developers comment to it though

**Comment by bkamins** [comment]
@nalimilan - this should be good to have a look at (especially corner cases in the tests).

File: test/indexing.jl [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module TestIndexing
 
-using Test, DataFrames
+using Test, DataFrames. OffsetArrays
```
Comment (bkamins) [inline test/indexing.jl:3]:
```suggestion
using Test, DataFrames, OffsetArrays
```

**Comment by bkamins** [comment]
I had to split the tests into a separate file as OffsetArrays.jl does not run on Julia 1.0 (due to its dependency).

File: test/indexing.jl [hunk 3]
```diff
@@ -2005,4 +2005,8 @@ end
     @test_throws BoundsError lastindex(gk, 0)
 end
 
+if VERSION >= v"1.5"
+    include("indexing_offest.jl")
```
Comment (bkamins) [inline test/indexing.jl:2009]:
```suggestion
    include("indexing_offset.jl")
```

**Comment by bkamins** [comment]
CI and coverage passes cleanly now.

File: NEWS.md [hunk 4]
```diff
@@ -6,6 +6,11 @@
 
 ## Bug fixes
 
+* DataFrames.jl now checks that passed columns are 1-based as this is a current
+  design assumption; `mapcols!` makes sure not to create columns being
+  `AbstractRange` consistently with other methods that add columns to a `DataFrame`
```
Comment (nalimilan) [inline NEWS.md:11]: This part is probably worth a separate bullet.

File: Project.toml [hunk 5]
```diff
@@ -45,9 +45,10 @@ DataStructures = "864edb3b-99cc-5e75-8d2d-829cb0a9cfe8"
...
 Dates = "ade2ca70-3891-5945-98fb-dc099432e06a"
 Logging = "56ddb016-857b-54e1-b83d-db4d58db5568"
+OffsetArrays = "6fe1bfb0-de20-5000-8ca7-80f57d26f881"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 Unitful = "1986cc42-f94f-5a68-af5c-568840ba703d"
 
 [targets]
-test = ["Combinatorics", "DataStructures", "DataValues", "Dates", "Logging", "Random", "Test", "Unitful"]
+test = ["Combinatorics", "DataStructures", "DataValues", "Dates", "Logging", "OffsetArrays", "Random", "Test", "Unitful"]
```
Comment (nalimilan) [inline Project.toml:54]: 92 chars limit ? :-p

File: src/abstractdataframe/iteration.jl [hunk 6]
```diff
@@ -454,7 +454,17 @@ function mapcols!(f::Union{Function, Type}, df::DataFrame)
     if len_min != len_max
         throw(DimensionMismatch("lengths of returned vectors must be identical"))
     end
-    _columns(df) .= vs
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl:457]: What was the problem with this syntax? Compilation overhead?

File: src/dataframe/dataframe.jl [hunk 7]
```diff
@@ -369,8 +377,26 @@ DataFrame(column_eltypes::AbstractVector{<:Type}, cnames::AbstractVector{<:Abstr
...
+# 4. if length of the vector is changed that the index of the `DataFrame`
+#    is adjusted appropriately
 _columns(df::DataFrame) = getfield(df, :columns)
 
+_onebased_check_error() =
+    throw(ArgumentError("Currently DataFrames.jl supports only columns " *
+                        "that use 1-based indexing"))
+_onebased_check_error(i, fic) =
+    throw(ArgumentError("Currently DataFrames.jl supports only " *
+                        "columns that use 1-based indexing and " *
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:397]:
```suggestion
                        "columns that use 1-based indexing, but " *
```

File: src/dataframe/dataframe.jl [hunk 8]
```diff
@@ -369,8 +377,26 @@ DataFrame(column_eltypes::AbstractVector{<:Type}, cnames::AbstractVector{<:Abstr
 ##############################################################################
 
 index(df::DataFrame) = getfield(df, :colindex)
+
+# this function grants the access to the internal storage of columns of the
+# `DataFrame` and its use is unsafe. If the returned vector is mutated then
+# make sure that:
+# 1. `AbstractRange` columns are not added to a `DataFrame`
+# 2. all inserted columns use 1-based indexing
+# 3. that after several mutating operations on the vector are performed
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:386]:
```suggestion
# 3. after several mutating operations on the vector are performed
```

File: src/dataframe/dataframe.jl [hunk 9]
```diff
@@ -205,6 +208,11 @@ struct DataFrame <: AbstractDataFrame
             end
         end
 
+        for (i, col) in enumerate(columns)
+            fic = firstindex(col)
+            fic != 1 && _onebased_check_error(i, fic)
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:213]: Given that this pattern happens in lots of places and that we don't care about performance when an error happens, I would suggest changing `_onebased_check_error` to take `col`:
```suggestion
            firstindex(col) != 1 && _onebased_check_error(i, col)
```

File: src/abstractdataframe/iteration.jl [hunk 10]
```diff
@@ -454,7 +454,17 @@ function mapcols!(f::Union{Function, Type}, df::DataFrame)
...
+
+    for (i, col) in enumerate(vs)
+        fic = firstindex(col)
+        fic != 1 && _onebased_check_error(i, fic)
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl:460]: Given that this pattern happens in lots of places and that we don't care about performance when an error happens, I would suggest changing `_onebased_check_error` to take `col`:
```suggestion
        firstindex(col) != 1 && _onebased_check_error(i, col)
```

File: test/indexing_offset.jl [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    DataFrames._check_consistency(df)
+
+    @test_throws ArgumentError df.b = ov1
+    @test_throws ArgumentError insertcols!(df, :b => ov1)
+    @test_throws DimensionMismatch df[!, :b] .= ov1
+
+    # this is consequence of the fact that OffsetArrays wrap AbstractRange in this case
+    @test_throws ErrorException df[:, :a] = ov1
+
+    # this inconsistency is the consequence how setindex! for vector is deifned in Base
```
Comment (nalimilan) [inline test/indexing_offset.jl:28]:
```suggestion
    # this inconsistency is the consequence how setindex! for vector is defined in Base
```

**Comment by yurivish** [comment]
@bkamins I just learned about the function `Base.require_one_based_indexing` – are you aware of it? It's mentioned in the devdocs [here](https://docs.julialang.org/en/v1/devdocs/offset-arrays/) and I'm not sure if it's sanctioned for use in user code, but the devdocs are written as if it is.

It is defined here: https://github.com/JuliaLang/julia/blob/ae53238c45a0cd6dafc6e121f4daaa93143bf627/base/abstractarray.jl#L103

**Comment by bkamins** [comment]
@yurivish - thank you for looking into this. I preferred custom checks for the following reasons (although the function seems to be a part of "official" API even if it is not exported):
* it is not available in Julia 1.0 which we support
* we can throw custom error message if we handle the check ourselves
* it is, easy to make this check

File: src/abstractdataframe/iteration.jl (hunk 6 above)
Comment (bkamins) [reply src/abstractdataframe/iteration.jl:457]: yes - I try to remove broadcasting in the internal code to reduce latency a bit (also I thought that the syntax might seem too magic for casual reader).

File: src/abstractdataframe/iteration.jl (hunk 10 above)
Comment (bkamins) [reply src/abstractdataframe/iteration.jl:460]: But `col` is a vector. What would you then print in the error message?

File: NEWS.md (hunk 4 above)
Comment (bkamins) [reply NEWS.md:11]: fixed

File: Project.toml (hunk 5 above)
Comment (bkamins) [reply Project.toml:54]: fixed

File: src/abstractdataframe/iteration.jl (hunk 6 above)
Comment (bkamins) [reply src/abstractdataframe/iteration.jl:457]: The difference is:

current release:
```
julia> using DataFrames

julia> df = DataFrame(a=1)
1×1 DataFrame
 Row │ a     
     │ Int64 
─────┼───────
   1 │     1

julia> @time mapcols!(length, df)
  0.003640 seconds (2.75 k allocations: 146.932 KiB)
1×1 DataFrame
 Row │ a     
     │ Int64 
─────┼───────
   1 │     1
```
this PR:
```
julia> using DataFrames

julia> df = DataFrame(a=1)
1×1 DataFrame
 Row │ a     
     │ Int64 
─────┼───────
   1 │     1

julia> @time mapcols!(length, df)
  0.000015 seconds (4 allocations: 256 bytes)
1×1 DataFrame
 Row │ a     
     │ Int64 
─────┼───────
   1 │     1
```
so it is not a big deal (but we have to iterate `vs` anyway to check `firstindex`).

File: src/dataframe/dataframe.jl [hunk 12]
```diff
@@ -369,8 +377,26 @@ DataFrame(column_eltypes::AbstractVector{<:Type}, cnames::AbstractVector{<:Abstr
...
+#    is adjusted appropriately
 _columns(df::DataFrame) = getfield(df, :columns)
 
+_onebased_check_error() =
+    throw(ArgumentError("Currently DataFrames.jl supports only columns " *
+                        "that use 1-based indexing"))
+_onebased_check_error(i, fic) =
+    throw(ArgumentError("Currently DataFrames.jl supports only " *
+                        "columns that use 1-based indexing, but " *
+                        "column $i has starting index equal to $fic"))
```
Comment (bkamins) [inline src/dataframe/dataframe.jl:398]:
```suggestion
_onebased_check_error(i, col) =
    throw(ArgumentError("Currently DataFrames.jl supports only " *
                        "columns that use 1-based indexing, but " *
                        "column $i has starting index equal to $(firstindex(col))"))
```

File: src/dataframe/dataframe.jl [hunk 13]
```diff
@@ -390,6 +416,12 @@ corrupt_msg(df::DataFrame, i::Integer) =
 
 function _check_consistency(df::DataFrame)
     cols, idx = _columns(df), index(df)
+
+    for (i, col) in enumerate(cols)
+        fic = firstindex(col)
+        fic != 1 && _onebased_check_error(i, fic)
```
Comment (bkamins) [inline src/dataframe/dataframe.jl:422]:
```suggestion
        firstindex(col) != 1 && _onebased_check_error(i, col)
```

File: src/abstractdataframe/iteration.jl (hunk 6 above)
Comment (nalimilan) [reply src/abstractdataframe/iteration.jl:457]: Wow. I wouldn't have expected such a large difference in the number of allocations.

**Review by nalimilan — APPROVED** [review]

**Comment by nalimilan** [comment]
Mmm, is the CI failure related to recent changes in Tables?

File: src/abstractdataframe/iteration.jl (hunk 6 above)
Comment (bkamins) [reply src/abstractdataframe/iteration.jl:457]: broadcasting is heavy (when doing the PRs that Tim Holy prompted I will review the whole codebase against using broadcasting). Of course here the difference is one-time and not that big.

**Comment by bkamins** [comment]
Yes, I will rebase and merge after CI passes.

**Comment by bkamins** [comment]
Thank you!
